nix repl: Print which variables are just loaded

kstrafe opened the pull request:
When we run `nix repl nixpkgs` we get "Added 6 variables". This is not useful as it doesn't tell us which variables the flake has exported to our global repl scope.

This patch prints the name of each variable that was just loaded.

Github issue: 11404

kstrafe commented on src/libcmd/repl.cc:
```diff
@@ -790,6 +806,21 @@ void NixRepl::addAttrsToScope(Value & attrs)
+
+    const int max_print = 10;
+    int count = 0;
+    for (auto & i : *attrs.attrs()) {
```
I'm not sure I'm not familiar with the API and thought that iteration would always be deterministic. I'll use this function instead as it seems more appropriate.

kstrafe replied:
Changed to use lexicographicOrder

kstrafe commented on src/libcmd/repl.cc:
```diff
@@ -470,6 +472,10 @@ ProcessLineResult NixRepl::processLine(std::string line)
         loadFlake(arg);
     }
 
+    else if (command == ":ll" || command == ":last-flake") {
```
ll = `last loaded`. The name `:show-scope` would be misleading I believe, since we're not actually showing the global scope, just the items that were most recently loaded. `:last-scope` or `:ls` might be a bit too close to the program `ls` and can cause confusion.
`:ll` applies to anything that loads an attr set into global scope, that is `:a`, `:l`, `:lf`, and repl cmdline args.

roberth commented:
Hi @kstrafe, how's it going?

Do you plan to continue to work on this, or do you prefer that someone else takes over?

kstrafe commented on tests/functional/repl.sh:
```diff
@@ -153,6 +153,49 @@ foo + baz
+- 7
+- 8
+- 9
+... And 3 other variables, use :ll to see these
```
Now doing comma-separated. We should ideally escape keys containing `,` and quotes and so on. Do we have an already existing function for this purpose? I could not find one.

kstrafe commented on src/libcmd/repl.cc:
```diff
@@ -790,6 +806,21 @@ void NixRepl::addAttrsToScope(Value & attrs)
+    for (auto & i : *attrs.attrs()) {
+        count += 1;
+        if (count > max_print) {
+            notice("... And %1% other variables, use :ll to see these", attrs.attrs()->size() - max_print);
```
Changed

kstrafe commented on src/libcmd/repl.cc:
```diff
@@ -790,6 +806,21 @@ void NixRepl::addAttrsToScope(Value & attrs)
+    lastLoaded = attrs;
+
+    const int max_print = 10;
+    int count = 0;
```
Done

kstrafe commented:
> Hi @kstrafe, how's it going?
> 
> Do you plan to continue to work on this, or do you prefer that someone else takes over?

Hey, I've just come back to this. I've added your suggestions and fixed up the tests. They aren't very easy to debug. I've added dumps to files and `echo Run vimdiff <x> <y> to see differences`. Would it be good to commit this (separate PR)? To rebuild and run the tests I've been using
```
meson setup --reconfigure
ninjaBuildPhase
meson test repl
```
Is there a more practical way? It's quite slow.

Anyhow, we now show all loaded variables on a single line. I have not handled escaping (for keys containing commas and spaces), but I believe that's quite rare so probably not very important to handle.

Is there any way to test the pager as well? I'd like to test `:ll` but have not figured out a way to do this.

kstrafe commented:
@roberth mentioning you since your profile mentioned it.

abdalazizrashid commented:
What is the state of this? What need be done to get it merged?

roberth commented on src/libcmd/repl.cc:
```diff
@@ -791,6 +807,27 @@ void NixRepl::addAttrsToScope(Value & attrs)
+        if (counter > 0)
+            loaded += ", ";
+
+        loaded += state->symbols[i->name];
```
`printIdentifier` would be nice to use here.

kstrafe replied:
Indeed, that was the one I was looking for 😀

roberth commented on src/libcmd/repl.cc:
```diff
@@ -791,6 +807,27 @@ void NixRepl::addAttrsToScope(Value & attrs)
+        counter += 1;
+    }
+
+    notice("- %1%", loaded);
```
Not sure what the dash is for. I'd expect a dash per variable, but it's already a comma separated list
```suggestion
    notice("%1%", loaded);
```
```suggestion
    notice("  %1%", loaded);
```

kstrafe replied:
To distinguish it from fixed outputs, I'll just remove it.

roberth reviewed (commented):
Good idea.
In addition to the other comments,
- this is worth adding a release note for
- could you add it to the `:help` command?

Mic92 commented:
Adding this to our triage, so it hopefully gets another review.

kstrafe commented:
> Adding this to our triage, so it hopefully gets another review.

Any updates?

tomberek commented on src/libcmd/repl.cc:
```diff
@@ -791,6 +808,27 @@ void NixRepl::addAttrsToScope(Value & attrs)
+
+    lastLoaded = attrs;
+
+    const int max_print = 10;
```
Perhaps something like 20-25? There are a some cases in the flake world where it will be more than 10.

kstrafe replied:
In that case the pager can be used. Is the team in agreement on this count? Alternatively, this can be changed in the future. I'd like to get this change in soon because it has been in limbo for a long while.

tomberek replied:
yes, I think 20 would be a good user experience

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2025-05-04-nix-team-meeting-minutes-230/65206/1

tomberek commented:
Pushed a rebase. (let me know if I messed something up). Otherwise the PR looks good. I'm interested in increasing the max_print count to 20.

Mic92 commented:
@tomberek could you also add a short release note snippet for it?

kstrafe commented:
> Pushed a rebase. (let me know if I messed something up). Otherwise the PR looks good. I'm interested in increasing the max_print count to 20.

`reloadFiles` is not part of this commit. I'll force push new changes rebased on to the top of master, with the `max_print` set to 20.

kstrafe commented:
New commits pushed. Limit increased to 20. Tests modified to reflect this.

roberth commented:
Thanks!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-2-30-0-released/66449/1
